src/function/mtk.jl

```diff
@@ -7,101 +7,31 @@ AutoModelingToolkit() = AutoModelingToolkit(false, false)
 
 function instantiate_function(f, x, adtype::AutoModelingToolkit, p, num_cons = 0)
     p = isnothing(p) ? SciMLBase.NullParameters() : p
-    sys = ModelingToolkit.modelingtoolkitize(OptimizationProblem(f, x, p))
 
-    hess_prototype, cons_jac_prototype, cons_hess_prototype = nothing, nothing, nothing
+    sys = ModelingToolkit.modelingtoolkitize(OptimizationProblem(f, x, p), num_cons = num_cons)
+    sys = ModelingToolkit.structural_simplify(sys)
+    f = OptimizationProblem(sys, x, p, grad = true, hess = true, obj_sparse = adtype.obj_sparse, cons_j = true, cons_h = true, cons_sparse = adtype.cons_sparse).f
 
-    if f.grad === nothing
-        grad_oop, grad_iip = ModelingToolkit.generate_gradient(sys, expression = Val{false})
-        grad(J, u) = (grad_iip(J, u, p); J)
-    else
-        grad = (G, θ, args...) -> f.grad(G, θ, p, args...)
-    end
+    grad = (G, θ, args...) -> f.grad(G, θ, p, args...)
```

Comment:
why put a closure? It's just `f.grad`. In fact, don't we just return `f`?

Reply:
The dispatch is defined differently there, it assumes the 2 arg dispatch to be the oop method, the symbolic version passes through https://github.com/SciML/Optimization.jl/blob/master/src/function/function.jl hence it also gets this closure defined

Reply:
I see